adr/0002-generate-extension-metadata-in-deployment-module.adoc

```diff
@@ -0,0 +1,117 @@
+
+Although it's not a serious problem, it would also be nice to allow people to use `CapabilityBuildItems` to register capabilities.
+At the moment, if authors do that, the information https://quarkus.io/guides/capabilities#capabilitybuilditem[will not be available to the dev tools], so we need to have lots of warnings in our documentation telling them not to.
+
+== Decision Drivers
+
+* Anything we do should minimise work for extension authors
+* Anything we do should minimise breaking changes, especially since many quarkiverse extensions will be using back-level versions of the Quarkus extension plugin
+* We should think carefully about the right information layout, so that we get the right information in the right long-term place, and only need to do this disruptive re-arrangement once
+
+== Considered options
+
+- *Manual creation of metadata*. This is tedious for users and error-prone; we could not assume any manually-managed metadata was complete. It's also annoyingly non-DRY; we'd be asking extension owners to write the same information down in two places (their code, and a metadata file). It does have the advantage that it's non-breaking.
+- *Running builds twice*. Because the circular dependency is a soft one, it works perfectly well to run a build to generate the deployment module's files, and then run a second build to pull those into the runtime module's metadata generation. Doing this would be a simple, non-breaking, change to release scripts. However, it is error-prone (if scripts are incorrectly run once-only, we just miss metadata), lengthens build times, breaks the Develocity by introducing hidden cross-module dependencies, and is generally Just Wrong.
+- *Doing nothing*. This is cheap, and non-breaking. But it means we miss an opportunity to take advantage of the fact that, at build time, we know a lot about what extensions can do and would like to make that information available to Quarkus users.
+- *Move (most) metadata generation to the deployment module* We can eliminate the circular dependency by not generating information which isn't particularly related to the runtime module in the runtime module. If the deployment module 'hosts' the `quarkus-extension.yaml`, it can harvest information from the extension's build steps and build items. This is architecturally clean, unlocks a range of capabilities in the future, and is highly disruptive in the present.
+
+== Proposal
+
+Moving metadata generation to the 'right' place is the architecturally correct thing to do.
+
+image::images/0002-resolution.svg[the proposed resolution]
+Source: https://excalidraw.com/#json=JMERR95gRufcmnUpWy6A_,Qd9efiR40LR3wv82CxyLKg
+
+So how do we minimise the impact?
+What needs to happen for it to work?
+
+* The registry reads `quarkus-extension.yaml` from the deployment model, not the runtime module
```

Comment:
```suggestion
* The registry reads `quarkus-extension.yaml` from the deployment module, not the runtime module
```

Reply:
I'm wondering whether we should also publish the yaml as its own Maven artifact with a classifier for these use-cases (in addition to bundling it in the deployment module).

Reply:
Resolving binaries for use-cases limited to metadata-only processing seems like an overkill.

Reply:
That's a good point. I guess there's no harm in publishing it as its own thing, is there? And it maybe has a nice (tiny) side-effect of exposing some of the quarkus-build-time-intelligence into maven.

Reply:
If we want to bundle the yaml in the deployment artifact then it would have to be generated as part of the deployment module build and if we want to publish it as its own artifact we have two options:
1. publish it from the deployment module (possibly with a classifier)
2. introduce a third (metadata) module that would depend on the deployment module, which would extract the yaml from the deployment module and publish it under `<quarkus-extension-artifactId>-metadata` artifactId (for example)

Reply:
We should probably also confirm we want it in YAML format. The platform descriptors are still in JSON format, which is essentially the same type of metadata.

Reply:
I wonder if that should be a separate change/consideration? I don't think there's a strong reason to do both changes at the same time.

Reply:
... but I agree using both json and yaml for similar information is slightly ugly.

Reply:
fwiw, we used yaml here as intended for easier human reading. 
was even before we did platform.json to be honest. if I could go back and change things I would just have picked one format either json or yaml not both. 

about additional archive, I get the point of being nice to look it up for external tools but it means duplication and yet another location to go look.

Reply:
Just to clarify, there wouldn't be an additional archive. It'd be yaml/json artifact. The advantage would be to not resolve and unpack binaries when just yaml/json is the target.